File: src/xuezh/core/clock.py
from __future__ import annotations

import os
import re
from datetime import datetime, timedelta, timezone

_DURATION_RE = re.compile(r"^(?P<n>[0-9]+)(?P<unit>[dh])$")


def now_utc() -> datetime:
    """Return the current time in UTC.

    Tests can override via `XUEZH_TEST_NOW_ISO` to make time-based logic deterministic.
    """
    override = os.environ.get("XUEZH_TEST_NOW_ISO")
    if override:
        return parse_utc_iso(override)
    return datetime.now(timezone.utc)


def parse_utc_iso(s: str) -> datetime:
    """Parse an ISO-8601 timestamp into a UTC datetime.

    Accepts either an explicit offset (e.g. `...+00:00`) or `Z`.
    """
    if s.endswith("Z"):
        s = f"{s[:-1]}+00:00"
    dt = datetime.fromisoformat(s)
    if dt.tzinfo is None:
        raise ValueError("Timestamp must be timezone-aware (include +00:00 or Z)")
    return dt.astimezone(timezone.utc)


def parse_duration(s: str) -> timedelta:
    """Parse `Nd` or `Nh` into a timedelta."""
    m = _DURATION_RE.match(s)
    if not m:
        raise ValueError(f"Invalid duration: {s!r} (expected Nd or Nh, e.g. 30d, 24h)")
    n = int(m.group("n"))
    unit = m.group("unit")
    if unit == "d":
        return timedelta(days=n)
    return timedelta(hours=n)

File: internal/xuezh/clock/clock.go
package clock

import (
	"errors"
	"os"
	"regexp"
	"strconv"
	"time"
)

var durationRe = regexp.MustCompile(`^(?P<n>[0-9]+)(?P<unit>[dh])$`)

func NowUTC() (time.Time, error) {
	if override := os.Getenv("XUEZH_TEST_NOW_ISO"); override != "" {
		return ParseUTCISO(override)
	}
	return time.Now().UTC(), nil
}

func ParseUTCISO(s string) (time.Time, error) {
	if len(s) >= 1 && s[len(s)-1] == 'Z' {
		s = s[:len(s)-1] + "+00:00"
	}
	if t, err := time.Parse(time.RFC3339Nano, s); err == nil {
		if t.Location() == time.UTC {
			return t.UTC(), nil
		}
		return t.UTC(), nil
	}
	if t, err := time.Parse(time.RFC3339, s); err == nil {
		return t.UTC(), nil
	}
	return time.Time{}, errors.New("timestamp must be timezone-aware (include +00:00 or Z)")
}

func ParseDuration(s string) (time.Duration, error) {
	m := durationRe.FindStringSubmatch(s)
	if m == nil {
		return 0, errors.New("invalid duration: expected Nd or Nh, e.g. 30d, 24h")
	}
	var nStr, unit string
	for i, name := range durationRe.SubexpNames() {
		if name == "n" {
			nStr = m[i]
		}
		if name == "unit" {
			unit = m[i]
		}
	}
	n, err := strconv.Atoi(nStr)
	if err != nil {
		return 0, errors.New("invalid duration: expected Nd or Nh, e.g. 30d, 24h")
	}
	switch unit {
	case "d":
		return time.Duration(n) * 24 * time.Hour, nil
	case "h":
		return time.Duration(n) * time.Hour, nil
	}
	return 0, errors.New("invalid duration: expected Nd or Nh, e.g. 30d, 24h")
}

func FormatISO(t time.Time) string {
	t = t.UTC().Truncate(time.Microsecond)
	if t.Nanosecond() == 0 {
		return t.Format("2006-01-02T15:04:05-07:00")
	}
	return t.Format("2006-01-02T15:04:05.000000-07:00")
}
